### ISSUE TITLE
Can't use await in templates

### ISSUE THREAD

**Issue by dawkot** [nim-lang/Nim#12743]
### Example
```nim
import asyncdispatch

template templ = await sleepAsync 0

proc prc {.async.} = templ
```
or
```nim
import asyncdispatch

template templ {.dirty.} = await sleepAsync 0

proc prc {.async.} = templ
```

### Current Output
```
/usercode/in.nim(5, 12) template/generic instantiation of `async` from here
/playground/nim/lib/pure/asyncmacro.nim(448, 10) Error: Await only available within .async
```

### Expected Output
No errors

### Additional Information
```
$ nim -v
Nim Compiler Version 1.0.2
```

**Comment by hlaaftana** [comment]
A template with no return type is set to return `void`, which is a typed expression. You need to do `template templ: untyped`.

**Comment by zedeus** [comment]
Known issue, #3885 says to use `yield` instead. Also #12085 but it's blocked by #12372

**Comment by dawkot** [comment]
Funny, it works on JS backend
```nim
import asyncjs, dom

template t =
  await newPromise do(resolve: proc()):
    discard setTimeout(resolve, 3000)

proc p {.async.} =
  t
  echo "Hello"

discard p()
```

**Comment by alehander92** [comment]
i should fix this this week , sorry

**Comment by metagn** [comment]
This was fixed by #12085 and a test was added
